### PR TITLE
test(client): preserve BigInt precision in relationJoins queries

### LINKED ISSUE #29010: BigInt precision loss in nested relations with relationJoins preview feature
## Bug description

When using driver adapters (e.g., `@prisma/adapter-pg`) with `previewFeatures = ["relationJoins"]`, BigInt values in nested relations lose precision silently.

The root cause is that `relationJoins` uses PostgreSQL lateral joins with JSON aggregation to fetch related data. When this JSON is parsed with `JSON.parse()`, integers exceeding `Number.MAX_SAFE_INTEGER` (2^53 - 1) are converted to JavaScript Numbers, causing precision loss.

### Schema

```prisma
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["driverAdapters", "relationJoins"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model User {
  id    BigInt     @id @default(autoincrement())
  files UserFile[]
}

model UserFile {
  id     String @id @default(uuid())
  userId BigInt
  user   User   @relation(fields: [userId], references: [id])
}
```

### Test code

```typescript
import { PrismaClient } from '@prisma/client'
import { PrismaPg } from '@prisma/adapter-pg'
import { Pool } from 'pg'

const pool = new Pool({ connectionString: process.env.DATABASE_URL })
const adapter = new PrismaPg(pool)
const prisma = new PrismaClient({ adapter })

async function main() {
  // Create test data
  const user = await prisma.user.create({ data: {} })
  console.log('Created user with id:', user.id) // e.g., 312590077454712834n

  const file = await prisma.userFile.create({
    data: { userId: user.id }
  })

  // Fetch with include (triggers relationJoins lateral join)
  const result = await prisma.userFile.findUnique({
    where: { id: file.id },
    include: { user: true }
  })

  console.log('Direct userId:', result.userId)      // 312590077454712834n (correct)
  console.log('Nested user.id:', result.user.id)    // 312590077454712832n (WRONG!)
  console.log('Match:', result.userId === result.user.id) // false!
}

main()
```

## Expected behavior

`result.userId` should equal `result.user.id` - the BigInt value should be preserved exactly.

## Actual behavior

The nested relation's BigInt value is corrupted:
- `result.userId` = `312590077454712834n` (correct, from direct column)
- `result.user.id` = `312590077454712832n` (corrupted, from JSON-parsed relation)

The type remains `bigint`, but the value is silently wrong.

## Environment

- Prisma version: 6.x / 7.x with `relationJoins` preview feature
- Database: PostgreSQL
- Driver adapter: `@prisma/adapter-pg` (likely affects all driver adapters)
- OS: Any

## Fix

PR: https://github.com/prisma/prisma/pull/29009

The fix adds a `safeJsonParse()` function that preserves large integers by converting them to strings before parsing. The existing data mapper then correctly converts these strings to BigInt.

### PR DESCRIPTION
**Pull request by polaz**
## Summary

This PR adds a regression test for a critical bug where BigInt values lose precision when using driver adapters (like `@prisma/adapter-pg`) with the `relationJoins` preview feature enabled.

Fixes #29010

### The Problem

When `previewFeatures = ["relationJoins"]` is enabled, Prisma uses lateral joins to fetch related data. The related records are returned as JSON-aggregated columns. When parsing this JSON data, standard `JSON.parse()` converts large integers (exceeding `Number.MAX_SAFE_INTEGER`) to JavaScript Numbers, silently corrupting the values.

**Example corruption:**
- Original value: `312590077454712834`
- After JSON.parse: `312590077454712800` (precision lost!)

### The Fix

The fix is implemented at the SQL level in `prisma-engines`:
- Postgres: cast BigInt to text in `JSONB_BUILD_OBJECT` (prisma/prisma-engines#5745)
- MySQL + CockroachDB: cast BigInt/INT8 values to string in JSON builders (prisma/prisma-engines#5752)

BigInt columns are cast to string before JSON aggregation, so values arrive as strings and are correctly parsed without precision loss.

**Changes in this PR:**
- Added regression test in `packages/client/tests/functional/issues/29010-bigint-precision-relation-joins/`
- Minor type safety improvement (explicit `as Value` cast) in `data-mapper.ts`

## Reproduction

<details>
<summary>Schema</summary>

```prisma
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["driverAdapters", "relationJoins"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model User {
  id    BigInt     @id @default(autoincrement())
  files UserFile[]
}

model UserFile {
  id     String @id @default(uuid())
  userId BigInt
  user   User   @relation(fields: [userId], references: [id])
}
```

</details>

<details>
<summary>Test Code</summary>

```typescript
// Create test data
const user = await prisma.user.create({ data: {} })
const file = await prisma.userFile.create({ data: { userId: user.id } })

// Fetch with include (triggers relationJoins)
const result = await prisma.userFile.findUnique({
  where: { id: file.id },
  include: { user: true }
})

// Before fix: result.userId !== result.user.id (BigInt corrupted in nested relation)
// After fix: result.userId === result.user.id (correct)
```

</details>

## Test Plan

- [x] Regression test added for BigInt precision in relationJoins queries
- [x] Test covers PostgreSQL, CockroachDB, and MySQL (requires updated Wasm artifacts from prisma-engines#5752)
- [x] Tests nested relationJoins queries

## Breaking Changes

None. This is a bugfix that preserves intended behavior.

## Release Notes

* **Bug Fixes**
  * Improved JSON data parsing and type handling in data mapping operations.

* **Tests**
  * Added comprehensive test coverage for BigInt precision with relationJoins across PostgreSQL, CockroachDB, and MySQL database providers.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by polaz** [comment]
## Performance Fix

I've pushed a commit that addresses the 97.68% performance regression flagged by CodSpeed.

### Root Cause
The original implementation ran an expensive regex replacement on **every** JSON parse call, even when no large integers existed in the data.

### Solution
Added a fast-path check that skips the regex replacement for the common case:

```typescript
const HAS_LARGE_INT = /\d{16}/

export function safeJsonParse(json: string): unknown {
  // Fast path: skip expensive regex if no 16+ digit sequences exist
  if (!HAS_LARGE_INT.test(json)) {
    return JSON.parse(json)
  }
  // Slow path: only when large integers might be present
  // ...
}
```

### Why This Works
- The simple `/\d{16}/.test()` check returns early for most strings
- Only JSON containing potential large integers (16+ digits) triggers the replacement
- The BigInt precision fix still applies when needed (relationJoins with BigInt foreign keys)

All existing tests pass. Waiting for CI to confirm the performance improvement.

File: packages/client-engine-runtime/src/utils.ts [hunk 1]
```diff
@@ -126,3 +126,56 @@ export function safeJsonStringify(obj: unknown): string {
     return val
   })
 }
+
+/**
+ * Regular expression to match JSON integer values that may lose precision when
+ * parsed as JavaScript Number. Matches integers with more than 15 significant digits.
```
[the remark this answers is not in the document]
Comment (polaz) [reply packages/client-engine-runtime/src/utils.ts]: Fixed. Updated the comment to say "16 or more significant digits" for consistency with the regex pattern.

File: packages/client-engine-runtime/src/utils.test.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  })
+
+  test('handles nested BigInt values', () => {
+    const obj = {
+      user: {
+        id: BigInt('312590077454712834'),
+        files: [{ fileId: BigInt('412590077454712834') }],
+      },
+    }
+    const result = JSON.parse(safeJsonStringify(obj))
```
[the remark this answers is not in the document]
Comment (polaz) [reply packages/client-engine-runtime/src/utils.test.ts]: Fixed. Changed to use `safeJsonParse` for proper round-trip consistency.

File: packages/client-engine-runtime/src/utils.ts [hunk 3]
```diff
@@ -126,3 +126,56 @@ export function safeJsonStringify(obj: unknown): string {
...
+ *
+ * Uses lookahead (?=...) to avoid consuming the suffix character, allowing
+ * consecutive matches in arrays like `[123..., 456...]`.
+ *
+ * Capturing groups:
+ * - Group 1: The prefix (`:`, `[`, or `,` with optional whitespace)
+ * - Group 2: The integer value (16+ digits, optionally negative)
+ * - Lookahead asserts suffix (`,`, `}`, or `]`) without consuming it
+ */
+const LARGE_INT_PATTERN = /(:\s*|[,\[]\s*)(-?\d{16,})(?=\s*[,}\]])/g
```
[the remark this answers is not in the document]
Comment (polaz) [reply packages/client-engine-runtime/src/utils.ts]: Fixed. Updated the regex to use a two-alternative pattern that first matches string literals (to skip them) before matching large integers. Added capturing group around the string literal pattern and updated the replacer to return the match unchanged when the string literal group is defined.

Also added two new tests:
- `preserves large integers inside string values unchanged`
- `handles escaped quotes in strings with large integers`

File: packages/client-engine-runtime/src/utils.ts [hunk 4]
```diff
@@ -126,3 +126,59 @@ export function safeJsonStringify(obj: unknown): string {
...
+  // Numbers with 16+ digits may exceed MAX_SAFE_INTEGER and lose precision.
+  // The regex has two alternatives: string literals (group 1) and integers (groups 2-3).
+  // When a string literal is matched, return it unchanged to avoid corruption.
+  const safeJson = json.replace(LARGE_INT_PATTERN, (match, stringLiteral, prefix, num) => {
+    if (stringLiteral !== undefined) {
+      return match
+    }
+    return `${prefix}"${num}"`
+  })
+  return JSON.parse(safeJson)
```
[the remark this answers is not in the document]
Comment (polaz) [reply packages/client-engine-runtime/src/utils.ts]: Fixed. Added `TOP_LEVEL_LARGE_INT` pattern (`/^-?\d{16,}$/`) to handle bare scalar integers. When the JSON is just a large integer, it's returned as a string directly without regex replacement.

This handles edge cases like JSON columns storing numeric scalars.

File: packages/client-engine-runtime/src/utils.test.ts [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    const unsafeParsed = JSON.parse(json)
+    expect(unsafeParsed.userId).not.toBe(312590077454712834n) // Not BigInt, it's a Number
+    // The exact corruption value varies by JS engine, just verify it's corrupted
+    expect(unsafeParsed.userId.toString()).not.toBe('312590077454712834')
+
+    // safeJsonParse preserves the value as string
+    const safeParsed = safeJsonParse(json)
+    expect((safeParsed as { userId: string }).userId).toBe('312590077454712834') // Exact value preserved
+  })
+})
```
[the remark this answers is not in the document]
Comment (polaz) [reply packages/client-engine-runtime/src/utils.test.ts]: Added. New test `handles top-level large integer JSON values` covers both positive and negative cases:
```typescript
expect(safeJsonParse('312590077454712834')).toBe('312590077454712834')
expect(safeJsonParse('-312590077454712834')).toBe('-312590077454712834')
```

File: packages/client-engine-runtime/src/utils.test.ts [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    const obj = {
+      user: {
+        id: BigInt('312590077454712834'),
+        files: [{ fileId: BigInt('412590077454712834') }],
+      },
+    }
+    const result = safeJsonParse(safeJsonStringify(obj))
+    expect(result.user.id).toBe('312590077454712834')
+    expect(result.user.files[0].fileId).toBe('412590077454712834')
+  })
```
[the remark this answers is not in the document]
Comment (polaz) [reply packages/client-engine-runtime/src/utils.test.ts]: Fixed. Added type assertion for the round-trip result to ensure TypeScript compatibility.

File: packages/client-engine-runtime/src/utils.test.ts [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+    const unsafeParsed = JSON.parse(json)
+    expect(unsafeParsed.userId).not.toBe(312590077454712834n) // Not BigInt, it's a Number
+    // The exact corruption value varies by JS engine, just verify it's corrupted
+    expect(unsafeParsed.userId.toString()).not.toBe('312590077454712834')
+
+    // safeJsonParse preserves the value as string
+    const safeParsed = safeJsonParse(json)
+    expect((safeParsed as { userId: string }).userId).toBe('312590077454712834') // Exact value preserved
+  })
+})
```
Comment (aqrln) [inline packages/client-engine-runtime/src/utils.test.ts]: this looks superfluous

**Comment by aqrln** [comment]
Thank you for the pull request.

I'm not sure about this approach to be honest. I would prefer if we added the casts to the SQL queries so that those values are strings in the JSON in the first place.

**Comment by polaz** [comment]
> Thank you for the pull request.
> 
> I'm not sure about this approach to be honest. I would prefer if we added the casts to the SQL queries so that those values are strings in the JSON in the first place.

i'm not happy with the approach as well - but had to highlight the problem as encountered in actual app

**Comment by polaz** [comment]
> Thank you for the pull request.
> 
> I'm not sure about this approach to be honest. I would prefer if we added the casts to the SQL queries so that those values are strings in the JSON in the first place.

Please take a look at https://github.com/prisma/prisma-engines/pull/5745, if i got it correctly

**Comment by aqrln** [comment]
@polaz I merged your engines PR, feel free to add a regression test in `packages/client/tests/functional` after the WebAssembly modules are built and updated in this repo by the bot.

**Comment by polaz** [comment]
Replying to review threads:

**Thread 1 (aqrln - "this looks superfluous"):** Removed the redundant test in commit 1cfaff0e9.

**Thread 2 (CodeRabbit - comment accuracy):** This test was for `safeJsonParse` which has been removed. The comment is no longer relevant.

**Threads 3-5 (Copilot - missing safeJsonParse):** The PR description was outdated. The fix is now implemented at the SQL level in prisma-engines (https://github.com/prisma/prisma-engines/pull/5745), where BigInt values are cast to `::text` in `JSONB_BUILD_OBJECT` before JavaScript parses them. 

The `safeJsonParse` workaround was initially added but then **removed** because the SQL-level fix handles precision at the database level. The functional tests will pass once the Wasm modules are updated with the prisma-engines fix.

File: packages/client-engine-runtime/src/utils.test.ts
Comment (aqrln) [inline packages/client-engine-runtime/src/utils.test.ts:1]: This is a test suite for the function that's now been removed, this file should be gone too

File: packages/client/tests/functional/issues/29010-bigint-precision-relation-joins/tests.ts [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+  },
+  {
+    skip: (skip, conf) => {
+      skip(!conf.previewFeatures?.includes('relationJoins'), 'this test is only for relation joins')
+    },
+    optOut: {
+      from: ['mongodb', 'sqlite', 'sqlserver'],
+      reason: 'relationJoins not supported',
+    },
+  },
```
Comment (aqrln) [inline packages/client/tests/functional/issues/29010-bigint-precision-relation-joins/tests.ts:69]: non blocking: I think there's no harm in running this test without the relation joins too, it's a valuable test to have generally

**Review by aqrln — APPROVED** [review]
LGTM, thanks so much!

**Comment by aqrln** [comment]
@polaz tests now show that the bug still exists in MySQL and, more surprisingly, in CockroachDB.

MySQL needs a similar fix in its implementation of `visit_json_build_object` method. I'm not fully sure what's going on with CockroachDB, but I assume it's probably because the native type is called `INT` and not `BIGINT` there, so the `(Some(TypeFamily::Int), Some("BIGINT"))` pattern doesn't match it.

Would you be interested in fixing those two as well? If not, then let's exclude them from this test for now and create the corresponding issues.

**Comment by polaz** [comment]
I addressed the MySQL/CockroachDB precision issue in prisma-engines: https://github.com/prisma/prisma-engines/pull/5752 (casts BigInt/INT8 to text in JSON builders, plus tests). Once Wasm updates land, the functional tests should pass for those providers as well.

File: packages/client-engine-runtime/src/interpreter/data-mapper.ts [hunk 9]
```diff
@@ -68,7 +68,7 @@ function mapArrayOrObject(
   if (typeof data === 'string') {
     let decodedData: Value
     try {
-      decodedData = JSON.parse(data)
+      decodedData = JSON.parse(data) as Value
```
Comment (jacek-prisma) [inline packages/client-engine-runtime/src/interpreter/data-mapper.ts]: what's the reason for changing this?

Comment (polaz) [reply packages/client-engine-runtime/src/interpreter/data-mapper.ts]: You're right, it was superfluous — `Value` is just `unknown` and `JSON.parse` returns `any`, so the cast adds nothing. Removed.
